extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js

```diff
@@ -105,23 +122,191 @@ export class AmpInstallServiceWorker extends AMP.BaseElement {
     iframe.src = this.iframeSrc_;
     this.element.appendChild(iframe);
   }
+
+  /** @private */
+  maybeInstallUrlRewrite_() {
+    // Only run rewrite in single-doc environment.
+    if (!this.getAmpDoc().isSingleDoc()) {
+      return;
+    }
+
+    const ampdoc = this.getAmpDoc();
+    const win = this.win;
+    const winUrl = parseUrl(win.location.href);
+
+    // Cannot install url rewrite on proxy.
+    if (isProxyOrigin(winUrl)) {
+      return;
+    }
+
+    // Read the url-rewrite config.
+    const urlMatch = this.element.getAttribute(
+        'data-no-service-worker-fallback-url-match');
+    let shellUrl = this.element.getAttribute(
+        'data-no-service-worker-fallback-shell-url');
+    if (!urlMatch && !shellUrl) {
+      return;
+    }
+
+    // Check the url-rewrite config is valid.
+    user().assert(urlMatch && shellUrl,
+        'Both, "%s" and "%s" must be specified for url-rewrite',
+        'data-no-service-worker-fallback-url-match',
+        'data-no-service-worker-fallback-shell-url');
+    shellUrl = removeFragment(shellUrl);
+    let urlMatchExpr;
+    try {
+      urlMatchExpr = new RegExp(urlMatch);
+    } catch (e) {
+      throw user().createError(
+          'Invalid "data-no-service-worker-fallback-url-match" expression', e);
+    }
+    user().assert(winUrl.origin == parseUrl(shellUrl).origin,
+        'Shell origin "%s" must be the same as source origin "%s"',
+        shellUrl, winUrl.href);
+
+    // Install URL rewriter.
+    this.urlRewriter_ = new UrlRewriter_(ampdoc, urlMatchExpr, shellUrl);
+
+    // Cache shell.
+    this.waitToPreloadShell_(shellUrl);
+  }
+
+  /**
+   * @param {string} shellUrl
+   * @return {!Promise}
+   * @private
+   */
+  waitToPreloadShell_(shellUrl) {
+    // Ensure that document is loaded and visible first.
+    const whenReady = this.loadPromise(this.win);
+    const whenVisible = viewerForDoc(this.getAmpDoc()).whenFirstVisible();
+    return Promise.all([whenReady, whenVisible]).then(() => {
+      this.deferMutate(() => this.preloadShell_(shellUrl));
+    });
+  }
+
+  /**
+   * @param {string} shellUrl
+   * @private
+   */
+  preloadShell_(shellUrl) {
+    const win = this.win;
+
+    // Preload the shell by via an iframe with `#preload` fragment.
+    const iframe = win.document.createElement('iframe');
+    iframe.id = 'i-amp-shell-preload';
+    iframe.setAttribute('src', shellUrl + '#preload');
```

Comment:
@dvoytenko To make the preload effective, the `shellUrl` must have cache headers right? Only then when a user clicks match link, the shell will be immediately served from cache?